### PR TITLE
deepFreezr

### PR DESCRIPTION
**Pull request by davesnx**
Hello @carlesba 

This PR aims for begin the approach of doing a Deep Freeze with the freezr method, it's just a WIP, we can discuss or work together with this.

First of all, as we talked :tongue:, I did:
- Abstract the utils methods on src
- Abstract the helpers on the spects (Just for avoid repetition of thousand of lines of code: created the `mockFrozenBuilder`)
- Improve [freezrs](https://github.com/carlesba/freezr/blob/master/src/freezers.js), with the detection of Array or Object, adding there the `deepFreeze` fn .
- Adding the proper tests for the `Object` deepFreeze.

TODO: 
- More unit tests for the object. (Check attributes/lengths/add functions)
- Create tests for the array 
- Could be nice to have, the callstack for check the recursion!

### CODE REVIEW

File: src/freezers.js [hunk 1]
```diff
@@ -1,7 +1,42 @@
...
+    Object.assign(o, keyIterator(source, (value, key) => recursiveFreeze(value, key)))
+  } else {
+    o[key] = source
+  }
+
+  if (Array.isArray(source)) {
+    return source.map(value => recursiveFreeze(value))
+  }
+
+  return freeze(o)
```
Comment (davesnx) [inline src/freezers.js]: I didn't find the way without having the `o` as a temp. I accomplish with freezing the childs of the source or at least the parent, not both. I don't know if there's a way more functional.

File: Readme.md [hunk 2]
```diff
@@ -79,5 +79,6 @@ frozenObject.delete('a') // will return another frozen object without 'a'
 
 ## To Do
 
-Create a deep parser.
```
Comment (carlesba) [inline Readme.md:82]: out of the scope in the PR, but appreciate it.

File: test/Freezr.spec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import {
+  buildMockObjectFrozen,
+  buildMockObjectDeepFrozen,
+  buildMockArrayFrozen,
+  // buildMockArrayDeepFrozen,
+  expectToBeImmutableAndThrowError
+} from './helpers'
```
Comment (carlesba) [inline test/Freezr.spec.js]: move builders to the test files, improves test readability.

File: test/Freezr.spec.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import {
```
Comment (carlesba) [inline test/Freezr.spec.js]: File should be called: freezers.spec.js as is testing freezers.js file

File: test/Freezr.spec.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    const { frozen } = buildMockObjectFrozen()
+    expectToBeImmutableAndThrowError(frozen)
+  })
+
+  it('returns an immutable array', () => {
+    const { frozen } = buildMockArrayFrozen()
+    expectToBeImmutableAndThrowError(frozen)
+  })
+
+  describe('deepFreezr', function () {
```
Comment (carlesba) [inline test/Freezr.spec.js]: test name: deepFreeze.
it shouldn't be nested as it's being exported in freezer.js

File: test/Freezr.spec.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  buildMockObjectFrozen,
+  buildMockObjectDeepFrozen,
+  buildMockArrayFrozen,
+  // buildMockArrayDeepFrozen,
+  expectToBeImmutableAndThrowError
+} from './helpers'
+
+describe('freezr', function () {
+  it('returns an immutable object', () => {
+    const { frozen } = buildMockObjectFrozen()
```
Comment (carlesba) [inline test/Freezr.spec.js]: the point of the test is to freeze an Object and check if it's frozen so it'll be better to see the source object to improve readability.

File: test/helpers.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import expect from 'expect'
```
Comment (carlesba) [inline test/helpers.js:1]: move this methods back to the previous files please.

File: test/Freezr.spec.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import {
+  buildMockObjectFrozen,
+  buildMockObjectDeepFrozen,
+  buildMockArrayFrozen,
+  // buildMockArrayDeepFrozen,
+  expectToBeImmutableAndThrowError
+} from './helpers'
+
```
Comment (carlesba) [inline test/Freezr.spec.js]: Tested methods should be imported in the test file.

File: test/Freezr.spec.js (hunk 4 above)
Comment (davesnx) [inline test/Freezr.spec.js]: sure!

File: test/Freezr.spec.js (hunk 5 above)
Comment (davesnx) [inline test/Freezr.spec.js]: :+1:

File: test/Freezr.spec.js (hunk 8 above)
Comment (davesnx) [inline test/Freezr.spec.js]: It will, when I'm testing something that isn't a mocked structure!

**Comment by davesnx** [comment]
I'm against add readability on the tests adding 20 lines of repeated code creating the same or look alike mocks.

Just in case the `freezers.spec.js` needs this 4 mocks:

``` js
export const buildMockObjectFrozen = () => mockFrozenBuilder(freeze, createMockObject())
export const buildMockArrayFrozen = () => mockFrozenBuilder(freeze, createMockArray())
export const buildMockObjectDeepFrozen = () => mockFrozenBuilder(deepFreeze, createMockNestedObject())
export const buildMockArrayDeepFrozen = () => mockFrozenBuilder(deepFreeze, createMockNestedArray())
```

If I actually copy/pasting them, it's 68 lines of code :-1:

File: src/freezers.js [hunk 9]
```diff
@@ -1,7 +1,46 @@
...
+    return new FrozenObject(source)
+  } else if (Array.isArray(source)) {
+    return new FrozenArray(source)
+  } else {
+    return source
+  }
+}
+
+const recursiveObject = (source, key) => {
+  let o = {}
```
Comment (carlesba) [inline src/freezers.js:23]:
``` js
import FrozenObject from './FrozenObject'
import FrozenArray from './FrozenArray'

const isPlainObj = (value) => {
  return value && (value.constructor === Object || value.constructor === undefined)
}

export const deepFreeze = (source) => {
  return freeze(recursiveFreeze(source))
}

export const freeze = (source) => {
  if (isPlainObj(source)) {
    return new FrozenObject(source)
  } else if (Array.isArray(source)) {
    return new FrozenArray(source)
  } else {
    return source
  }
}

const recursiveFreeze = (source) => {
  if (isPlainObj(source)) {
    return keyIterator(source, (v, k) => deepFreeze(v))
  } else if (Array.isArray(source)) {
    return source.map((v) => deepFreeze(v))
  } else {
    return source
  }
}

const keyIterator = (obj, fn) => {
  return Object.keys(obj).reduce((acc, key) => {
    acc[key] = fn(obj[key], key)
    return acc
  }, {})
}
```

**Comment by davesnx** [comment]
ty

**Pull request closed without merging**